runtime/compiler/control/JitDump.cpp

```diff
@@ -153,29 +177,69 @@ jitDumpRecompileWithTracing(J9VMThread *vmThread, J9Method *ramMethod, TR::Compi
    // created before the compile is issued will use the trace log we provide to initialize IL tracing.
-   TR_OptimizationPlan *plan = TR_OptimizationPlan::alloc(optLevel);
-   if (NULL == plan)
+   TR_OptimizationPlan *plan = NULL;
+#if defined(J9VM_OPT_JITSERVER)
+   // JITServer will create a new optimization plan on the client
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() != JITServer::SERVER)
+#endif
       {
-      j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
-      return;
-      }
+      plan = TR_OptimizationPlan::alloc(optLevel);
+      if (NULL == plan)
+         {
+         j9nls_printf(PORTLIB, J9NLS_INFO | J9NLS_STDERR, J9NLS_DMP_JIT_OPTIMIZATION_PLAN);
+         return;
+         }
 
-   plan->setInsertInstrumentation(isProfilingCompile);
-   plan->setLogCompilation(jitdumpFile);
+      plan->setInsertInstrumentation(isProfilingCompile);
+      plan->setLogCompilation(jitdumpFile);
+      }
 
    trfprintf(jitdumpFile, "<recompilation>\n");
 
    // Set the VM state of the crashed thread so the diagnostic thread can use consume it
    compInfo->setVMStateOfCrashedThread(vmThread->omrVMThread->vmState);
 
    J9::JitDumpMethodDetails details(ramMethod, optionsFromOriginalCompile, isAOTBody);
-   auto rc = compilationOK;
    auto queued = false;
-   compInfo->compileMethod(vmThread, details, oldStartPC, TR_no, &rc, &queued, plan);
+   auto rc = compilationOK;
+#if defined(J9VM_OPT_JITSERVER)
+   if (compInfo->getPersistentInfo()->getRemoteCompilationMode() == JITServer::SERVER)
+      {
+      // Inform the client that the compilation thread has crashed and send it
+      // a pointer to the JitDump file, so that it can request
+      // a JitDump recompilation from the server with an updated plan and method details.
+      TR_MethodToBeCompiled *entry = TR::compInfoPT->getMethodBeingCompiled();
+      JITServer::ServerStream *stream = entry->_stream;
+      stream->write(JITServer::MessageType::compilationThreadCrashed, jitdumpFile);
+      stream->read<JITServer::Void>();
+         {
+         // Add an entry to the compilation queue using the current stream,
+         // and immediatelly set its method details as JitDump method details,
+         // so that only the diagnostic thread can compile it.
+         OMR::CriticalSection compilationMonitorLock(compInfo->getCompilationMonitor());
+
+         compInfo->requeueOutOfProcessEntry(entry);
+         entry->_methodDetails = TR::IlGeneratorMethodDetails::clone(entry->_methodDetailsStorage, details);
+         queued = true;
+         }
+      // Wait for the recompilation to finish
+      entry->acquireSlotMonitor(vmThread);
+      entry->getMonitor()->wait();
+      entry->releaseSlotMonitor(vmThread);
+      }
+   else
+#endif
+      {
+      compInfo->compileMethod(vmThread, details, oldStartPC, TR_no, &rc, &queued, plan);
+      }
 
    trfprintf(jitdumpFile, "</recompilation rc=%d queued=%d>\n", rc, queued);
```

Comment:
One last comment: this line that shows JIT dump recompilation started, appears after the recompilation has finished for JITServer. We need to move this line up before the crashed thread starts waiting.

Reply:
This line will appear after recompilation finished for non-JITServer as well, should I move that too?

Reply:
You are right because `compInfo->compileMethod` is a synchronous compilation. It also sort of makes sense because it also prints the return code. I looked at that `queued` variable and I assumed we wanted to print that the compilation has been queued and therefore it's about to start.
We can leave it as is, then.